### PR TITLE
Feat/add multiclient

### PR DESCRIPTION
**Pull request by kittylyst**
* Adds multiclient
* Refactors to allow report decoration by the client
  - This allows previous failure attempts to be recorded and sent up
  - Significant cleanup of the closures that actually do the report sending

### CODE REVIEW

File: api/src/main/java/com/redhat/insights/AbstractTopLevelReportBase.java [hunk 1]
```diff
@@ -133,7 +138,7 @@ public void generateReport(Filtering masking) {
         defaultHost = host.getHostName();
       }
     } catch (UnknownHostException e) {
-      logger.error("Unknown Host in lookup", e);
+      logger.error("Unknown Host in lookup, continuing with localhost", e);
```
Comment (ehsavoie) [inline api/src/main/java/com/redhat/insights/AbstractTopLevelReportBase.java:141]: shouldn't be at error level

File: jboss-cert-helper/main.go [hunk 2]
```diff
@@ -1,99 +1,99 @@
 package main
 
 import (
-  "errors"
-  "fmt"
-  "log"
-  "os"
-  "os/user"
-  "syscall"
+	"errors"
```
Comment (ehsavoie) [inline jboss-cert-helper/main.go:4]: Why has format changed ? This is not pertinent in this PR

File: api/src/main/java/com/redhat/insights/http/InsightsMultiClient.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  private final List<InsightsHttpClient> clients;
+
+  public InsightsMultiClient(InsightsLogger logger, List<InsightsHttpClient> clients) {
+    this.logger = logger;
+    this.clients = clients;
+  }
+
+  public InsightsMultiClient(InsightsLogger logger, InsightsHttpClient... clients) {
+    this.logger = logger;
+    this.clients = Arrays.asList(clients);
```
[the remark this answers is not in the document]
Comment (kittylyst) [reply api/src/main/java/com/redhat/insights/http/InsightsMultiClient.java:27]: Why?

File: api/src/main/java/com/redhat/insights/http/InsightsMultiClient.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      try {
+        if (previousExceptionMsg != "") {
+          report.decorate("client.exception", previousExceptionMsg);
+        }
+        client.sendInsightsReport(filename, report);
+        return;
+      } catch (InsightsException x) {
+        logger.debug("Client failed, trying next", x);
+        byte[] msgBytes = x.getMessage().getBytes(StandardCharsets.UTF_8);
+        previousExceptionMsg = Base64.getEncoder().encodeToString(msgBytes);
```
[the remark this answers is not in the document]
Comment (kittylyst) [reply api/src/main/java/com/redhat/insights/http/InsightsMultiClient.java]: Good catch. Fixed.

File: jboss-cert-helper/main.go (hunk 2 above)
Comment (jponge) [reply jboss-cert-helper/main.go:4]: `go fmt` is the way to go. So it should be tabs, not spaces 😉

Comment (kittylyst) [reply jboss-cert-helper/main.go:4]: go fmt appears is a no-op on my machine - so I think this is correct.

Comment (jponge) [reply jboss-cert-helper/main.go:4]: Yes, `go fmt` from `jboss-cert-helper` and you're good to go.

**Pull request merged**